hydroshare_jupyter_sync/lib/filesystem/fs_map.py

```diff
@@ -0,0 +1,217 @@
+        resource id's."""
+        return (
+            [d.name for d in self.fs_root.glob("*") if len(d.name) == 32 and d.is_dir()]
+            if self.fs_root
+            else []
+        )
+
+
+# Concrete implementations
+
+
+class LocalFSMap(FSMap, IEntityFSMap):
+    """Class representing the relationship between *local* HydroShare resources', resource files,
+    and resource MD5 Hashes."""
+
+    def __init__(self, fs_root: Union[str, Path]) -> Dict[ResourceId, FSResourceMap]:
+        super().__init__()
+        self.fs_root = Path(fs_root).expanduser().resolve()
+
+    # override
+    @classmethod
+    def create_map(cls, fs_root: Union[Path, str]) -> "LocalFSMap":
+        # create class instance
+        fs_map = cls(fs_root)
+
+        for resource in fs_map._get_resource_ids():
+            # create new instance of LocalFSResourceMap
+            fs_map[resource] = LocalFSResourceMap(fs_map.fs_root / resource)
+
+        return fs_map
+
+    def add_resource(self, resource_id: ResourceId) -> None:
+        """Create new LocalFSResourceMap and add to LocalFSMap instance. `resource_id` provided must
+        be direct child directory of `fs_root`."""
+        if resource_id not in self.data:
+            # create new local resource map
+            r_map = LocalFSResourceMap.from_resource_path(self.fs_root / resource_id)
+
+            # add local resource map to dictionary
+            self.data[resource_id] = r_map
+
+    def add_resource_file(
+        self, resource_id: ResourceId, relative_resource_file: Union[Path, str]
+    ) -> None:
+        if resource_id in self.data:
+            self.data[resource_id].add_file(relative_resource_file)
+
+    def update_resource_file(
+        self, resource_id: ResourceId, relative_resource_file: Union[Path, str]
+    ) -> None:
+        """Update a file in a LocalFSResourceMap instance for a given resource id."""
+        if resource_id in self.data:
+            self.data[resource_id].update_file(relative_resource_file)
+
+    def delete_resource_file(
+        self, resource_id: ResourceId, relative_resource_file: Union[Path, str]
+    ) -> None:
+        """Remove a file in a LocalFSResourceMap instance for a given resource id."""
+        if resource_id in self.data:
+            self.data[resource_id].delete_file(relative_resource_file)
+
+
+class RemoteFSMap(FSMap):
+    """Class representing the relationship between remote HydroShare resource's, resource files, and
+    resource MD5 Hashes."""
+
+    def __init__(
+        self, fs_root: Union[str, Path], hydroshare: HydroShare
+    ) -> Dict[ResourceId, FSResourceMap]:
+        super().__init__()
+        self.fs_root = Path(fs_root).expanduser().resolve()
+        self._hydroshare = hydroshare
+
+    # override
+    @classmethod
+    def create_map(
+        cls, fs_root: Union[Path, str], hydroshare: HydroShare
+    ) -> "RemoteFSMap":
```

Comment:
Is this a thing?  Returning a string of the class name to be returned?

Reply:
I'm having some deja-vu, feels like I've asked this before.

Reply:
Yeah haha. I should use [PEP 563](https://www.python.org/dev/peps/pep-0563/) to get around this. It is the way to provide type hints without the whole class having been interpreted. I think this changes in `3.10`, but as of now, type hint interpretation is not deferred to after the class has been interpreted.